lib/std/math.zig

```diff
@@ -1279,6 +1279,35 @@ test "lossyCast" {
     try testing.expect(lossyCast(u32, @as(f32, maxInt(u32))) == maxInt(u32));
 }
 
+/// Performs linear interpolation between *a* and *b* based on *t*.
+/// *t* must be in range 0.0 to 1.0. T must be a float type.
+///
+/// This does not guarantee returning *b* if *t* is 1 due to floating-point errors.
+/// This is monotonic.
+pub fn lerp(a: anytype, b: anytype, t: anytype) @TypeOf(a, b, t) {
+    const T = @TypeOf(a, b, t);
+
+    comptime if (!std.meta.trait.isFloat(T))
+        @compileError("T must be a float type");
+
+    assert(t >= 0 and t <= 1);
+    return @mulAdd(T, b - a, t, a);
```

Comment:
I know y'all have been digging into lots of different implementations, but I wanted to put forward my recommendation:
```zig
return if (t <= 0.5) a+(b-a)*t else b-(b-a)*(1.0-t);
```

This includes a branch, but it is also precise at both endpoints and additionally is monotonic (`lerp(a, b, t) >= lerp(a, b, t')` if `t = t'`) and for nearby `t` values the branch should be well-predicted.


I think it's better to prioritize these numerical properties over the performance, since they mean this operation will compose better with surrounding code. Of course, if a user knows that that they don't need precision at the endpoints, they are free to write `(b - a)*t + a` themselves.

Reply:
Any issues with changing that to `if (t <= 0.5) @mulAdd(t, b - a, a) else b - (b - a) * (1.0 - t)`?

Reply:
Yes indeed - `if (t <= 0.5) @mulAdd(t, b - a, a) else @mulAdd(t, a - b, b)` would be just fine too

Reply:
Perhaps this calls for two functions, `lerpAccurate`/`lerpFast`? `lerp` could then be selected depending on `FloatMode`?

Reply:
I'm sorry but I'm not entirely convinced, at least right now.
I use `lerp` in some deep places in my code where I could really use the best performance, which we can get quite easily here while still having a lot of precision.

> The formula `(b - a)*t + a` is pretty obvious

Not to me personally. If I wrote it often enough I could memorize it but generally I would look it up every time, or I could think it up from scratch, but at least for me it would take a lot longer than if I can just use `std.math.lerp`.
On top of that, what this std implementation provides is that it also uses `@mulAdd` for more precision and performance.
This is something most people just won't really think of if they write a quick `(b - a) * t + a` function.

> but it is also precise at both endpoints

It sounds like this will make it so that "This does not guarantee returning *b* if *t* is 1 due to floating-point errors." is no longer true.
But still for general use cases is the precision with `@mulAdd` added not enough?
I mean, for general use when do you actually need it to be this precise at both endpoints? Do you have an example or something?

> and additionally is monotonic

Is it not already monotonic in its current form? Is the documentation "This is monotonic." I included wrong?

In this case I would like to go back to the solution of having multiple implementations to choose from. So we would have one `lerp` for general use cases and then a `lossyLerp` or a `preciseLerp` for other cases depending on what that `lerp` is going to be.

> I'm not convinced a `lerpFast` is worth including in the stdlib.

But yeah, we should probably try to avoid doing something like that.

There are just really a lot of ways to do this. It's not easy.

Reply:
One area where lerp is very useful is gamedev. In gamedev people will not care about these tiny differences in precision, at least I wouldn't. In gamedev I would rather appreciate the performance. If you just use it for visual effects and stuff it's either not going to make a difference or matter, I believe. Maybe I'm wrong.

Do you have an example or can you name some other area where this kind of additional precision is necessary for lerp?

Reply:
I guess it depends on whether we want `std.math` to be general-purpose or ultra-high-performance.

My recommendation for extremely performance-sensitive applications would be to use a third-party lib or write the optimized version yourself. There's also nothing wrong with adding a comment for ultra-high-perf use cases:
```zig
/// Guarantees for finite a, b:
///    - lerp(a, b, 0.0) == a
///    - lerp(a, b, 1.0) == b
///    - lerp(a, b, t) >= lerp(a, b, t') for t >= t' and b => a
///    - lerp(a, b, t) <= lerp(a, b, t') for t >= t' and b <= a
///
/// If numerical imprecision is tolerable, a faster alternative is @mulAdd(T, b - a, t, a)
fn lerp(a: anytype, b: anytype, t: anytype) @TypeOf(a,b,t) {
```

That said, if we'd like to commit to providing "fast, but imprecise" functions in stdlib, maybe it's worth considering a separate `std.fastmath` namespace.

I don't have more input on this one, so I'll defer to @andrewrk to make the right call

Reply:
P.S. It looks like on x86-64 LLVM is smart enough to optimize this branch away and use `blendvpd` instead:
```asm
lerp:
        vaddsd  xmm3, xmm2, qword ptr [rip + .LCPI6_0]
        vcmpnlesd       xmm4, xmm2, qword ptr [rip + .LCPI6_1]
        vsubsd  xmm1, xmm1, xmm0
        vblendvpd       xmm2, xmm2, xmm3, xmm4
        vmulsd  xmm1, xmm1, xmm2
        vaddsd  xmm0, xmm1, xmm0
        ret
```

Reply:
> Do you have an example or can you name some other area where this kind of additional precision is necessary for lerp?

The main advantage of this alternative is that it is bounded by `[a,b]`

For example, `acos(lerp(a, b, t) / max(abs(a), abs(b)))` is never `nan` for the precise version (unless both a and b are huge, both >= std.math.floatMax(T) / 2). Using the imprecise version here results in `nan` sometimes (e.g. a = -0.5, b = 10000001)

If I have to write an equation with a restricted-domain function like this (e.g. sqrt, log, acos, etc.), then I'd have to clamp the output of lerp to avoid producing erroneous results.

The imprecise lerp is still a good choice in many situations of course, and I agree there's no one-size-fits-all here. Nonetheless, I think bounded-ness is a useful property to expect for "general-purpose" applications.